### PR TITLE
Pet applications

### LINKED ISSUE #3: User Stories 4-5

### PR DESCRIPTION
**Pull request by noahdurbin**
- section on applications show page to search for a pet
- each pet has a button to add pet to application
- tests for these features

### CODE REVIEW

File: app/controllers/applications_controller.rb [hunk 1]
```diff
@@ -1,6 +1,9 @@
 class ApplicationsController < ApplicationController
   def show
-    @applications = Application.find(params[:id])
+    @application = Application.find(params[:id])
+    return unless params[:pet_name]
```
Comment (JCL461437) [inline app/controllers/applications_controller.rb:4]: Interesting conditional use.

File: spec/features/applications/show_spec.rb [hunk 2]
```diff
@@ -24,5 +24,35 @@
...
+        visit "/applications/#{@application1.id}"
+
+        fill_in 'pet_name', with: 'Buddy'
+
+        click_button 'Submit'
+
+        expect(page).to have_content('Buddy')
+      end
+
+      it 'has a button and can add each pet to application' do
```
Comment (JCL461437) [inline spec/features/applications/show_spec.rb:44]: I think these tests work well as unit tests. Maybe we should look into unit feature tests in the future.

File: spec/features/applications/show_spec.rb [hunk 3]
```diff
@@ -4,14 +4,14 @@
   before(:each) do
     @shelter1 = Shelter.create!(foster_program: true, name: 'Whimsy Whiskers', city: 'Denver', rank: 42)
     @pet1 = @shelter1.pets.create!(name: 'Buddy', breed: 'Golden Retriever', age: 3, adoptable: true)
-    @application1 = Application.create!(name: 'John Doe', address: '123 Elm St, Denver, CO 12392', description: 'Looking for a friendly dog', status: 'Pending')
-    PetApplication.create!(pet: @pet1, application: @application1)
+    @application1 = Application.create!(name: 'John Doe', address: '123 Elm St', city: 'Denver', state: 'CO',
+                                        zip: '12345', description: 'Looking for a friendly dog', status: 'In Progress')
+    # PetApplication.create!(pet: @pet1, application: @application1)
```
Comment (AuroraTuring) [inline spec/features/applications/show_spec.rb]: Let's make sure when we merge in code, we don't leave unnecessary commented out stuff, general housekeeping.

File: app/views/applications/show.html.erb [hunk 4]
```diff
@@ -1,7 +1,22 @@
...
+
+<% if @application.status == "In Progress" %>
+  <%= "Add a Pet to this Application" %>
+  <%= form_with url: "/applications/#{@application.id}", method: :get, local: true do |form| %>
+    <br> <%= form.label :pet_name %>
+    <%= form.text_field :pet_name %>
+
+    <br> <%= form.submit "Submit" %>
+  <% end %>
+  <% if params[:pet_name] %>
```
Comment (JCL461437) [inline app/views/applications/show.html.erb:17]: I think this conditional was a good solution for determining whether we should display what we wanted in the form depending on whether the params we wanted were present. Good work Noah!

File: config/routes.rb [hunk 5]
```diff
@@ -5,6 +5,9 @@
   get '/applications/:id', to: 'applications#show'
   post '/applications', to: 'applications#create'
 
+  get '/pet_applications/new', to: 'pet_applications#new'
+  post 'applications/:application_id/pet_applications', to: 'pet_applications#create'
```
Comment (JCL461437) [inline config/routes.rb:9]: That route was confusing, great work!

**Review by JCL461437 — APPROVED** [review]

File: app/controllers/applications_controller.rb (hunk 1 above)
Comment (noahdurbin) [reply app/controllers/applications_controller.rb:4]: It's good to know how much functionality we can put into these controllers, I think i thought that we could only put in what it has access to, not realizing their could be conditional logic in the controller as well.

File: spec/features/applications/show_spec.rb (hunk 3 above)
Comment (noahdurbin) [reply spec/features/applications/show_spec.rb]: good point, thanks for catching that.

Comment (JCL461437) [reply spec/features/applications/show_spec.rb]: Good idea! I should of thought about that

File: config/routes.rb (hunk 5 above)
Comment (noahdurbin) [reply config/routes.rb:9]: I'm pretty sure that was all you

File: app/views/applications/show.html.erb [hunk 6]
```diff
@@ -1,7 +1,22 @@
...
-<br>Application Status: <%= @applications.status %></p>
+<br>Application Status: <%= @application.status %></p>
+
+<% if @application.status == "In Progress" %>
+  <%= "Add a Pet to this Application" %>
+  <%= form_with url: "/applications/#{@application.id}", method: :get, local: true do |form| %>
+    <br> <%= form.label :pet_name %>
+    <%= form.text_field :pet_name %>
+
+    <br> <%= form.submit "Submit" %>
```
Comment (AuroraTuring) [inline app/views/applications/show.html.erb]: It behaves more like a search bar, so I would suggest changing "submit" to "search".

File: app/controllers/applications_controller.rb (hunk 1 above)
Comment (JCL461437) [reply app/controllers/applications_controller.rb:4]: Exactly! It is nice to know that our logic knowledge as related to Ruby can help in constructing a logical data flow.

File: app/views/applications/show.html.erb (hunk 6 above)
Comment (noahdurbin) [reply app/views/applications/show.html.erb]: I thought of that too, in the user story it said submit so thats why I went with submit but I would be up for changing it to search

Comment (JCL461437) [reply app/views/applications/show.html.erb]: I think that we were following the descriptions listed in the user story for naming buttons and sections of the form

File: app/controllers/applications_controller.rb (hunk 1 above)
Comment (AuroraTuring) [reply app/controllers/applications_controller.rb:4]: `@pets = Pet.search(params[:pet_name]) if params[:pet_name]`

Returning early can make the code wonky later on.

**Comment by AuroraTuring** [comment]
Noah, thanks for implementing those two most recent changes!

**Review by AuroraTuring — APPROVED** [review]

**Pull request merged**